File: Algorithm/baekjoon/bj_4921.py
cnt = 0 
piece = {"1":["4","5"], "3":["4","5"], "4":["2","3"],"5":["8"],"6":["2","3"],"7":["8"],"8":["6","7"]}
while True:
    cnt += 1
    a = input()
    # 입력이 0인 경우
    if a == "0": break
    # 시작이 1, 끝이 2가 아니면 안됨
    if a[0] != "1" or a[-1] != "2": print(f"{cnt}. NOT"); continue
    # 일반적인 경우, piece의 리스트에 담긴 번호만 뒤에 붙을 수 있음
    for i in range(len(a)-1):
        if not (piece.get(a[i]) and a[i+1] in piece[a[i]]):
            print(f"{cnt}. NOT")
            break
    else :
        print(f"{cnt}. VALID")

File: Algorithm/baekjoon/java/src/bj_4921.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st ;
        // piece
        HashMap<String, List<String>> piece = new HashMap<>();

        piece.put("1", Arrays.asList("4","5"));
        piece.put("3", Arrays.asList("4","5"));
        piece.put("4", Arrays.asList("2","3"));
        piece.put("5", Arrays.asList("8"));
        piece.put("6", Arrays.asList("2","3"));
        piece.put("7", Arrays.asList("8"));
        piece.put("8", Arrays.asList("6","7"));

//        System.out.println(piece);

        int cnt = 0;
        while (true) {
            cnt ++;
            // num
            st = new StringTokenizer(br.readLine());
            String num = st.nextToken();
            int numLength = num.length();

            // 입력이 0 인 경우
            if (num.equals("0")) break;
            // 시작이 1, 끝이 2가 아닌 경우
            if ( num.charAt(0) != '1' || num.charAt(numLength-1) != '2' ) {
                System.out.printf("%s. NOT%n", cnt);
                continue;
            }
            // 일반적인 경우
            boolean check = true;
            for (int i=0; i<numLength-1; i++) {
                String n = String.valueOf(num.charAt(i));
                String nextN = String.valueOf(num.charAt(i+1));
                if (!piece.containsKey(n) || !piece.get(n).contains(nextN) ) {
                    System.out.printf("%s. NOT%n", cnt);
                    check = false;
                    break;
                }
            }
            if (check) {
                System.out.printf("%s. VALID%n",cnt);
            }
        }
    }
}
